### PR TITLE
Deduplication on paired reads using clumpify and BBtools deduplication and fragment length analysis

### PR DESCRIPTION
**Pull request by harmonbhasin**
#60 should be merged first before attempting to merge this branch as it is based on that branch.

Deduplication and fragment length analysis from bbtools has been added to the `hv_hits_putative_collapsed.tsv.gz` table.

### CODE REVIEW

**Comment by jeffkaufman** [comment]
> #60 should be merged first before attempting to merge this branch as it is based on that branch.

The way to do this is to set `remove_concat_gzipped` as the merge target for this PR.  Then this will show up as a delta on top of #60

File: modules/local/bbmerge/main.nf [hunk 1]
```diff
@@ -5,6 +5,7 @@ process BBMERGE {
     input:
         tuple val(sample), path(reads)
     output:
+        tuple val(sample), path("${sample}_bbmerge_merged.fastq.gz"), emit: merged
```
Comment (willbradshaw) [inline modules/local/bbmerge/main.nf:8]: @harmonbhasin what's the thinking behind outputting the merged read file twice (once as `merged` and once in `reads`)? If you need to access this file specifically in another process, you can just access `reads[1][0]`.

File: modules/local/collapseHV/main.nf [hunk 2]
```diff
@@ -5,6 +5,8 @@ process COLLAPSE_HV {
     memory "16.GB"
     input:
         path(hv_hits_filtered)
+        path(merged_bbmerge_paired_results)
```
Comment (willbradshaw) [inline modules/local/collapseHV/main.nf:8]: This module is really only solving one specific problem, namely the presence of discordant read pairs in the HV TSV. I'd suggest moving the new parts to a different module.

File: modules/local/summarizeBBMerge/main.nf [hunk 3]
```diff
@@ -0,0 +1,17 @@
+process SUMMARIZE_BBMERGE {
+    label "base"
+    label "single"
+    input:
+        tuple val(sample), path(merged)
+    output:
+        path("${sample}_bbmerge_merged_summary.txt")
+    script:
+        """
+        zcat ${merged} | awk '
```
Comment (willbradshaw) [inline modules/local/summarizeBBMerge/main.nf:10]: Can we also get the post-merge sequence? This isn't currently included in the HV table and would be good to have.

(Ditto the post-concat sequence for unmerged read pairs, though this is somewhat easier to derive from the paired sequences.)

Comment (willbradshaw) [reply modules/local/summarizeBBMerge/main.nf:10]: Also, what's your plan for the merge status?

File: modules/local/summarizeDedup/main.nf [hunk 4]
```diff
@@ -0,0 +1,25 @@
+process SUMMARIZE_DEDUP {
```
Comment (willbradshaw) [inline modules/local/summarizeDedup/main.nf:1]: Let's talk this one over in person.

File: subworkflows/local/taxonomy/main.nf [hunk 5]
```diff
@@ -32,15 +35,26 @@ workflow TAXONOMY {
         } else {
             subset_ch = SUBSET_READS_PAIRED(reads_ch, params.read_fraction)
         }
+
+         // Deduplicate reads (if applicable)
```
Comment (willbradshaw) [inline subworkflows/local/taxonomy/main.nf:39]: What's the thinking behind making this optional? Do you have a case in mind where we wouldn't want to do this?

**Comment by willbradshaw** [comment]
Independently from my other comments, it looks like this PR is trying to merge changes that are actually made in #60, which suggests something might have gone awry somewhere?

File: modules/local/bbmerge/main.nf (hunk 1 above)
Comment (harmonbhasin) [reply modules/local/bbmerge/main.nf:8]: @willbradshaw I ended up fixing this. The solution you suggested works, I just had to adjust the syntax.

File: modules/local/collapseHV/main.nf (hunk 2 above)
Comment (harmonbhasin) [reply modules/local/collapseHV/main.nf:8]: @willbradshaw done.

File: modules/local/summarizeBBMerge/main.nf (hunk 3 above)
Comment (harmonbhasin) [reply modules/local/summarizeBBMerge/main.nf:10]: @willbradshaw chatted about this during out meeting, but this was in collapseHV, which has now been moved to it's own module.

File: modules/local/summarizeDedup/main.nf (hunk 4 above)
Comment (harmonbhasin) [reply modules/local/summarizeDedup/main.nf:1]: @willbradshaw done.

**Review by willbradshaw — APPROVED** [review]

**Pull request merged**